File: Python/700. Search in a Binary Search Tree.py
'''
1.递归
2.迭代
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# class Solution:
#     def searchBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
#         if not root or root.val == val:
#             return root
#         if root.val < val:
#             return self.searchBST(root.right, val)
#         else:
#             return self.searchBST(root.left, val)

class Solution:
    def searchBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if not root:
            return root
        while (root):
            if root.val == val:
                return root
            elif root.val < val:
                root = root.right
            else:
                root = root.left
        return root

File: C++/700. Search in a Binary Search Tree.cpp
/*
1.µÝ¹é
2.µü´ú
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

 // class Solution {
 // public:
 //     TreeNode* searchBST(TreeNode* root, int val) {
 //         if (!root) return NULL;
 //         if (root->val == val) return root;
 //         if (root->val < val) return searchBST(root->right, val);
 //         else return searchBST(root->left, val);
 //     }
 // };

 // class Solution {
 // public:
 //     TreeNode* searchBST(TreeNode* root, int val) {
 //         stack<TreeNode*>st;
 //         st.push(root);
 //         while (!st.empty()){
 //             TreeNode* cur = st.top(); st.pop();
 //             if (cur != NULL){
 //                 if (cur->right) st.push(cur->right);
 //                 st.push(cur);
 //                 st.push(NULL);
 //                 if (cur->left) st.push(cur->left);
 //             }else{
 //                 TreeNode* cur = st.top(); st.pop();
 //                 if (cur->val < val) continue;
 //                 else if (cur->val == val) return cur;
 //                 else return NULL;
 //             }
 //         }
 //         return NULL;
 //     }
 // };

class Solution {
public:
    TreeNode* searchBST(TreeNode* root, int val) {
        if (!root) return root;
        while (root) {
            if (root->val == val) return root;
            else if (root->val < val) root = root->right;
            else root = root->left;
        }
        return NULL;
    }
};